Sources/BSON/BSONDecoder.swift

```diff
@@ -165,6 +154,33 @@ public class BSONDecoder {
         try self.decode(type, from: BSONDocument(fromBSON: data))
     }
 
+    /**
+     * Decodes a top-level value of the given type from the given BSON.
+     *
+     * - Parameter type: The type of the value to decode.
+     * - Parameter document: The BSON to decode from.
+     * - Returns: A value of the requested type.
+     * - Throws: `DecodingError` if any value throws an error during decoding.
+     */
+    internal func decode<T: Decodable>(_ type: T.Type, fromBSON bson: BSON) throws -> T {
+        // if the requested type is `BSON` we're done
+        if let b = bson as? T {
```

Comment:
if I remember correctly, we are using this method for the final conversion step right, from BSON to the user's requested type?
if so, it might make sense to also special-case when `bson.bsonValue` is an instance of `T`. 

in fact, I might be wrong but we actually might have to do that, because for all of our custom BSON types (e.g. `Binary`) we just throw an error if you actually try to use the generalized decoding implementation (called a few lines down), as those types (for now anyway) only are meant to work with encoders and decoders we own (BSON and now extJSON):
https://github.com/mongodb/swift-bson/blob/e0b0efbd7ea1175559a8e4f3cdf3d44de4813632/Sources/BSON/BSONValue.swift#L35

Reply:
Oh, nice catch. I think what would be even better would be to utilize the same code path that "interior" decoding would use by invoking `unbox` directly instead of `Type.init`. This is what [`JSONDecoder` does](https://github.com/apple/swift-corelibs-foundation/blob/master/Sources/Foundation/JSONEncoder.swift#L1163), for instance.

Reply:
`unbox` also has special handling if `T` is a `BSON` as well, so this `if let` can go away too.

Reply:
I think the change I just made should address this